File: morphir/sdk/core/src/morphir/sdk/Decimal.scala
package morphir.sdk
import java.math.{ BigDecimal => BigDec }

import morphir.sdk.Basics.{ Bool, Order }
import morphir.sdk.Maybe.Maybe

import scala.util.control.NonFatal

object Decimal extends DecimalModuleCompat {

  type Decimal = BigDecimal

  object Decimal {
    def apply(value: BigDec): Decimal                  = BigDecimal(value)
    def apply(value: scala.BigDecimal): Decimal        = value
    def apply(value: morphir.sdk.Float.Float): Decimal = BigDecimal.exact(value).bigDecimal
    def apply(value: morphir.sdk.Int.Int): Decimal     = BigDecimal(value).bigDecimal

  }

  @inline def apply(value: BigDec): Decimal                  = Decimal.apply(value)
  @inline def apply(value: scala.BigDecimal): Decimal        = Decimal.apply(value)
  @inline def apply(value: morphir.sdk.Float.Float): Decimal = Decimal.apply(value)
  @inline def apply(value: morphir.sdk.Int.Int): Decimal     = Decimal.apply(value)

  /** Absolute value (sets the sign as positive)
    */
  def abs(value: Decimal): Decimal = value.abs

  def add(a: Decimal)(b: Decimal): Decimal = a + b

  def bps(n: morphir.sdk.Int.Int): Decimal = Decimal(n * 0.0001)

  def compare(a: Decimal)(b: Decimal): Order =
    a.compareTo(b) match {
      case 0          => Order.EQ
      case n if n > 0 => Order.GT
      case _          => Order.LT
    }

  def div(a: Decimal)(b: Decimal): Maybe[Decimal] =
    if (b.compareTo(zero) == 0) Maybe.nothing
    else
      try {
        Maybe.just(a / b)
      } catch {
        case NonFatal(_) => Maybe.nothing
      }

  def divWithDefault(default: Decimal)(a: Decimal)(b: Decimal): Decimal =
    Maybe.withDefault(default)(div(a)(b))

  def eq(a: Decimal)(b: Decimal): Bool = a.compareTo(b) == 0

  def fromInt(value: morphir.sdk.Basics.Int): Decimal =
    Decimal(value)

  /** Converts a Float to a Decimal
    */
  def fromFloat(value: morphir.sdk.Float.Float): Decimal =
    Decimal(value)

  def fromString(value: morphir.sdk.String.String): Maybe[Decimal] =
    try {
      Maybe.just(new BigDec(value))
    } catch {
      case NonFatal(_) => Maybe.nothing
    }

  /** Converts an `Int` to a `Decimal`` that represents n hundreds.
    */
  def hundred(n: morphir.sdk.Int.Int): Decimal =
    Decimal(n * 100)

  def hundredth(n: morphir.sdk.Int.Int): Decimal =
    Decimal(n * 0.01)

  def gt(a: Decimal)(b: Decimal): Bool  = a.compareTo(b) > 0
  def gte(a: Decimal)(b: Decimal): Bool = a.compareTo(b) >= 0

  def lt(a: Decimal)(b: Decimal): Bool = a.compareTo(b) < 0

  def million(n: morphir.sdk.Int.Int): Decimal =
    Decimal(n * 1000000)

  def millionth(n: morphir.sdk.Int.Int): Decimal =
    Decimal(n * 0.000001)

  def mul(a: Decimal)(b: Decimal): Decimal = a * b

  @inline def ne(a: Decimal)(b: Decimal): Bool = neq(a)(b)
  def neq(a: Decimal)(b: Decimal): Bool        = a.compareTo(b) != 0

  def negate(value: Decimal): Decimal = -value

  def round(decimal: Decimal): Decimal = {
    val scale = decimal.scale
    decimal.setScale(scale, BigDecimal.RoundingMode.HALF_EVEN)
  }

  def shiftDecimalLeft(n: morphir.sdk.Int.Int)(value: Decimal): Decimal =
    value.bigDecimal.scaleByPowerOfTen(-n.intValue()) // TODO: When we align Int to Int this should settle in correctly

  def shiftDecimalRight(n: morphir.sdk.Int.Int)(value: Decimal): Decimal =
    value.bigDecimal.scaleByPowerOfTen(n.intValue()) // TODO: When we align Int to Int this should settle in correctly

  def sub(a: Decimal)(b: Decimal): Decimal = a - b

  def thousand(n: morphir.sdk.Int.Int): Decimal =
    Decimal(n * 1000)

  def toFloat(value: Decimal): morphir.sdk.Float.Float =
    morphir.sdk.Float.Float(value.toDouble)

  // TODO: Make sure the Elm call and this call return the same value
  def toString(value: Decimal): morphir.sdk.String.String = value.toString

  def _toString: Decimal => String = toString

  def truncate(decimal: Decimal): Decimal =
    decimal.setScale(0, BigDecimal.RoundingMode.DOWN)

  /** The number -1.
    */
  val minusOne: Decimal = BigDecimal.exact(-1).bigDecimal

  val one: Decimal = BigDec.ONE

  val zero: Decimal = BigDec.ZERO

}

File: pkg/models/stable/ir/sdk/decimal.go
package sdk

import (
	ir "github.com/finos/morphir-go/pkg/models/stable/ir"
)

// DecimalModuleName returns the module name for Morphir.SDK.Decimal
func DecimalModuleName() ir.ModuleName {
	return ir.PathFromString("Decimal")
}

// DecimalModuleSpec returns the module specification for Morphir.SDK.Decimal
func DecimalModuleSpec() ir.ModuleSpecification[ir.Unit] {
	return ir.NewModuleSpecification(
		decimalTypes(),
		decimalValues(),
		nil,
	)
}

// decimalTypes returns the type specifications for the Decimal module
func decimalTypes() []ir.ModuleSpecificationType[ir.Unit] {
	return []ir.ModuleSpecificationType[ir.Unit]{
		// Decimal type
		ir.ModuleSpecificationTypeFromParts[ir.Unit](
			ir.NameFromString("Decimal"),
			ir.NewDocumented(
				"Type that represents a decimal number.",
				ir.NewOpaqueTypeSpecification[ir.Unit](nil),
			),
		),
	}
}

// decimalValues returns the value specifications for the Decimal module
func decimalValues() []ir.ModuleSpecificationValue[ir.Unit] {
	return []ir.ModuleSpecificationValue[ir.Unit]{
		// Constants
		VSpec("zero", []VSpecInput{},
			DecimalType(),
			"The decimal value zero."),

		VSpec("one", []VSpecInput{},
			DecimalType(),
			"The decimal value one."),

		VSpec("minusOne", []VSpecInput{},
			DecimalType(),
			"The decimal value minus one."),

		// Constructors
		VSpec("fromInt", []VSpecInput{
			{Name: "value", Type: intType()},
		}, DecimalType(),
			"Convert an integer to a decimal."),

		VSpec("fromFloat", []VSpecInput{
			{Name: "value", Type: floatType()},
		}, DecimalType(),
			"Convert a float to a decimal."),

		VSpec("fromString", []VSpecInput{
			{Name: "str", Type: StringType()},
		}, MaybeType(DecimalType()),
			"Parse a string into a decimal."),

		// Scale helpers
		VSpec("hundred", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Multiply by 100."),

		VSpec("thousand", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Multiply by 1,000."),

		VSpec("million", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Multiply by 1,000,000."),

		VSpec("tenth", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Divide by 10."),

		VSpec("hundredth", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Divide by 100."),

		VSpec("thousandth", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Divide by 1,000."),

		VSpec("millionth", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Divide by 1,000,000."),

		VSpec("bps", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Convert basis points (1/10000)."),

		// Conversions
		VSpec("toString", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, StringType(),
			"Convert a decimal to a string."),

		VSpec("toFloat", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, floatType(),
			"Convert a decimal to a float."),

		// Arithmetic
		VSpec("add", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, DecimalType(),
			"Add two decimals."),

		VSpec("sub", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, DecimalType(),
			"Subtract two decimals."),

		VSpec("mul", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, DecimalType(),
			"Multiply two decimals."),

		VSpec("div", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, MaybeType(DecimalType()),
			"Divide two decimals, returning Nothing on division by zero."),

		VSpec("divWithDefault", []VSpecInput{
			{Name: "default", Type: DecimalType()},
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, DecimalType(),
			"Divide two decimals with a default value for division by zero."),

		VSpec("negate", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Negate a decimal."),

		VSpec("abs", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Get the absolute value of a decimal."),

		// Rounding
		VSpec("truncate", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, intType(),
			"Truncate a decimal to an integer."),

		VSpec("round", []VSpecInput{
			{Name: "value", Type: DecimalType()},
		}, intType(),
			"Round a decimal to the nearest integer."),

		// Comparison
		VSpec("gt", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, boolType(),
			"Greater than comparison."),

		VSpec("gte", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, boolType(),
			"Greater than or equal comparison."),

		VSpec("eq", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, boolType(),
			"Equality comparison."),

		VSpec("neq", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, boolType(),
			"Inequality comparison."),

		VSpec("lt", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, boolType(),
			"Less than comparison."),

		VSpec("lte", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, boolType(),
			"Less than or equal comparison."),

		VSpec("compare", []VSpecInput{
			{Name: "a", Type: DecimalType()},
			{Name: "b", Type: DecimalType()},
		}, OrderType(),
			"Compare two decimals, returning Order."),

		// Manipulation
		VSpec("shiftDecimalLeft", []VSpecInput{
			{Name: "positions", Type: intType()},
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Shift the decimal point left by the given number of positions."),

		VSpec("shiftDecimalRight", []VSpecInput{
			{Name: "positions", Type: intType()},
			{Name: "value", Type: DecimalType()},
		}, DecimalType(),
			"Shift the decimal point right by the given number of positions."),
	}
}

// DecimalType creates a Decimal type reference
func DecimalType() ir.Type[ir.Unit] {
	return ir.NewTypeReference(
		ir.Unit{},
		ToFQName(DecimalModuleName(), "Decimal"),
		nil,
	)
}
